src/devtools/useAtomsSnapshot.ts

```diff
@@ -1,36 +1,37 @@
-import { useContext } from 'react'
+import { useCallback, useContext } from 'react'
 import {
   SECRET_INTERNAL_getScopeContext as getScopeContext,
   SECRET_INTERNAL_useMutableSource as useMutableSource,
 } from 'jotai'
 import type { Atom, Scope } from '../core/atom'
-import {
-  getDebugStateAndAtoms,
-  subscribeDebugScopeContainer,
-} from '../core/Provider'
-import type { AtomState } from '../core/vanilla'
-// NOTE importing from '../core/Provider' is across bundles and actually copying code
+// NOTE importing from '../core/contexts' is across bundles and actually copying code
+import { isDevScopeContainer } from '../core/contexts'
+import { DEV_GET_ATOM_STATE, DEV_GET_MOUNTED } from '../core/store'
+import type { AtomState } from '../core/store'
 
 type AtomsSnapshot = Map<Atom<unknown>, unknown>
 
 export function useAtomsSnapshot(scope?: Scope): AtomsSnapshot {
   const ScopeContext = getScopeContext(scope)
-  const debugMutableSource = useContext(ScopeContext)[4]
+  const scopeContainer = useContext(ScopeContext)
 
-  if (debugMutableSource === undefined) {
+  if (!isDevScopeContainer(scopeContainer)) {
     throw Error('useAtomsSnapshot can only be used in dev mode.')
   }
 
-  const [state, atoms] = useMutableSource(
-    debugMutableSource,
-    getDebugStateAndAtoms,
-    subscribeDebugScopeContainer
+  const [store, , devMutableSource, devSubscribe] = scopeContainer
+
+  const atoms = useMutableSource(
+    devMutableSource,
+    // FIXME HACK creating new reference to force re-render
+    useCallback((devContainer) => [...devContainer.atoms], []),
+    devSubscribe
   )
 
   const atomToAtomValueTuples = atoms
-    .filter((atom) => !!state.m.get(atom))
+    .filter((atom) => !!store[DEV_GET_MOUNTED]?.(atom))
     .map<[Atom<unknown>, unknown]>((atom) => {
-      const atomState = state.a.get(atom) ?? ({} as AtomState)
+      const atomState = store[DEV_GET_ATOM_STATE]?.(atom) ?? ({} as AtomState)
       return [atom, atomState.e || atomState.p || atomState.w || atomState.v]
```

Comment:
@Thisen During refactoring, I noticed this might not be what we want because we don't want to save promises or errors in snapshot. Or, was this our intention? What do you think?

```suggestion
      return [atom, atomState.v]
```
Maybe just like this? As this PR is only refactor, would you create a dedicated PR please?

Reply:
@dai-shi I'll create a dedicated PR for it after you merged this